### DISCUSSION TITLE
[explore]users should not be allowed to save metric without sql expression

### DISCUSSION THREAD

**Discussion by junlincc** [apache/superset#19207] [Ideas]
currently, user can create a new metric in Edit dataset modal in Explore without sql expression, or without validating the new metric. 

the new metrics becomes available in metric popover dropdown for user.  once it is selected and run query, users receive a non human readable or non actionable error message, which leads to a bad user experience.. It also would become problematic once drag and drop query control is available. 

Proposed solution(shortterm): the modal should not allow user to save a metric without a VALID SQL expression. 
Long term - will require redesigning this modal as a whole.... 

https://user-images.githubusercontent.com/67837651/106103648-1f670a80-60f6-11eb-9637-8b67553c53df.mov

**Comment by zhaoyongjie** [comment]
@junlincc 
The existing architecture of Superset has no way to verify SQL snippet(metric/derived column), so we can only guarantee that the metric expression is not empty.
